Add logic to propagate SET LOCAL at xact start

jasonmp85 opened the pull request:
Partially addresses #1327 (which discusses _all_ config, not just `SET LOCAL`)

DESCRIPTION: Add logic to propagate SET LOCAL at xact start

# Tasks
  - [x] Add failing SQL test
  - [x] Add GUC to whitelist propagated settings (for now); do not allow PostgreSQL ones (for now)
  - [x] During transactions, detect matching `SET LOCAL` commands and…
  - [x] Push down to any already-open remote transactions, and…
  - [x] Append to a list of needed local settings for any further remote transactions, and…
  - [x] During any subsequent remote transactions, add the `SET` commands after the initial `BEGIN` (but in the same network hop)
  - [x] Show test passes

jasonmp85 commented:
# New Settings

FWIW I think the original suggestion we add some sort of GUC like `citus.gucs_synced_at_begin` isn't a bad one. In fact, it might not be a bad idea to straight-up only allow non-PostgreSQL GUCs in there to start (rather than delicately crafting a whitelist of what might be safe to send).

# MVP: Transaction Blocks

We'll want to keep track of any `SET LOCAL` things in the local transaction and push them down during any `BEGIN` to appropriate workers. Maybe "slow" (though I'm not actually sure of that, honestly), but correct/sufficient for a lot of stuff to become possible (including @mkurz's RLS policies).

# Additional Enhancements

OK, so in addition to explicit `SET LOCAL` commands within `BEGIN`, it appears there are two other places this could work…

## Functions

Declarations of functions can include configuration that is set/reset as the functions are entered/exited, respectively. Now that functions use transactions, propagating this should be straightforward.

## Multi-Statement Commands

This is a bit arcane, but it _is_ possible to use `SET LOCAL` without an explicit transaction block or function call. In particular, look at this…

```
psql 'SET LOCAL DateStyle TO SQL; SELECT now()' -c 'SHOW DateStyle;' postgres
┌───────────────────────────────┐
│              now              │
├───────────────────────────────┤
│ 04/24/2019 12:24:34.95311 MDT │
└───────────────────────────────┘
(1 row)

Time: 0.556 ms
┌───────────┐
│ DateStyle │
├───────────┤
│ ISO, MDY  │
└───────────┘
(1 row)

Time: 0.118 ms
```

The custom `DateStyle` appears to take affect for the `SELECT`. This means we could (in theory) preface SQL we generate (going to workers) with various `SET LOCAL` statements in order to force those single-SQL commands (not in transaction blocks) to run with the desired settings. As you can see above, the date style returns during the next transaction.

_BUT JASON ARE YOU SURE EACH `-c` COMMAND RUNS IN THE SAME CONNECTION?_

```
psql -c 'SELECT pg_backend_pid();' -c 'SELECT pg_backend_pid();' postgres
┌────────────────┐
│ pg_backend_pid │
├────────────────┤
│          16050 │
└────────────────┘
(1 row)

Time: 0.522 ms
┌────────────────┐
│ pg_backend_pid │
├────────────────┤
│          16050 │
└────────────────┘
(1 row)

Time: 0.103 ms
```

Yes, I am.

marcocitus commented:
> do not allow PostgreSQL ones

We should support propagating settings like enable_hashagg and work_mem. The fact that we don't propagate those is a frequent pain point.

mkurz commented:
So maybe a whitelist of allowed "PostgreSQL ones"?

hanefi commented:
> We should support propagating settings like enable_hashagg and work_mem

If we propagate changes to `work_mem` will that mean that the setting will be the same across the workers and the coordinator? AFAIK it may be better to have higher `work_mem` on the coordinator, and we should be careful in how we support this

marcocitus commented:
> If we propagate changes to work_mem will that mean that the setting will be the same across the workers and the coordinator? AFAIK it may be better to have higher work_mem on the coordinator, and we should be careful in how we support this

Yes, but only for the current transaction. You could use this to hint to the planner on each worker session that it should pick less memory hungry plan to prevent OOMs. work_mem does not affect the query plan on the coordinator in distributed queries.

sumedhpathak commented:
@jasonmp85 What is the status on this PR?

marcocitus commented on src/backend/distributed/commands/variableset.c:
```diff
@@ -0,0 +1,123 @@
+/*-------------------------------------------------------------------------
+ *
+ * vacuum.c
+ *    Commands for vacuuming distributed tables.
```
stale comment

jasonmp85 replied:
Yeah, I always copy-paste these when adding a new file; fixing…

marcocitus commented on src/backend/distributed/commands/utility_hook.c:
```diff
@@ -176,6 +176,20 @@ multi_ProcessUtility(PlannedStmt *pstmt,
+			ProcessVariableSetStmt(setStmt, queryString);
+		}
+
+		/* TODO: warn if not possible to propagate? */
```
I guess we probably shouldn't emit anything by default, since this potentially happens very frequently.

jasonmp85 replied:
I'll remove the `TODO`, but I'm _mostly_ worried about the case where a user is modifying a GUC that's part of the whitelist (admittedly not yet part of the PR when you looked at it) but _isn't_ doing so in a transaction so we don't propagate its effects. That seems like it might be counterintuitive enough for us to warn them.

marcocitus commented on src/backend/distributed/commands/variableset.c:
```diff
@@ -0,0 +1,123 @@
+			continue;
+		}
+
+		FinishRemoteTransactionVariableModify(connection);
```
We don't really care about the result very much except it not being an error. Could we do `ClearResults(connection, true)` here?

jasonmp85 replied:
Sure, I was mostly cargo-culting another function pair here.

jasonmp85 commented:
I left function comments for me to write tomorrow, but otherwise I've addressed your comments and added the whitelist GUC (for now I'm just parsing it out of a string on each `SET LOCAL`; that _could_ obviously be improved if you want, before merge).

What was the behavior you wanted for a whitelist: a single list with all GUCs that should be permitted, or a hardcoded one of some PostgreSQL GUCs we know can work, or…?

Let me know and I'll clean that up/document the functions/merge tomorrow, approval permitting.

mkurz commented:
Great to see you work on this!

I have one question: Will the example I gave in [this comment](https://github.com/citusdata/citus/issues/1327#issuecomment-384768838) work with this pull request?
I mean will the coordinator correctly figure out **to which worker send the query to when using RLS**?

E.g. when having a transaction like:
```sql
BEGIN;
SET LOCAL my_app.tenant_id=2; -- or SELECT set_config('my_app.tenant_id', '2', true);
SELECT * from events; -- fetches ONLY tenant_id with id 2 because of RLS rules
COMMIT; -- or ROLLBACK, doesn't matter for this example
```
Will the coordinator know about the RLS rules (only events with tenant_id = 2) and then, based on this restriction, send the query down to the worker that contains tenant with id 2, and then this worker will also set the GUC and apply the RLS?
I just want to make sure this works. If it does, this pull request is amazing :+1: and would also fix #1374 IMHO.

metdos commented:
Hi @marcocitus, is this on your plate now?

marcocitus commented on src/backend/distributed/commands/variableset.c:
```diff
@@ -0,0 +1,149 @@
+		MemoryContextSwitchTo(old_context);
+	}
+
+	/* asynchronously send SAVEPOINT */
```
stale comment

jasonmp85 replied:
Yeah, copypasta.

marcocitus commented on src/backend/distributed/commands/variableset.c:
```diff
@@ -0,0 +1,149 @@
+{
+	const bool raiseErrors = true;
+
+	ClearResults(connection, raiseErrors);
```
I'd probably just inline these Start/Finish functions.

jasonmp85 replied:
Sure; Done.

marcocitus commented on src/backend/distributed/transaction/transaction_management.c:
```diff
@@ -156,6 +140,25 @@ InitializeTransactionManagement(void)
+ * pre-existing coordinated transaction may be in progress./
+ */
+static void
+BeginCoordinatedTransaction(void)
```
Maybe move this back to not get `git blame`d

jasonmp85 replied:
Done.

jasonmp85 replied:
Actually, I'm going to leave this (though I'll keep it in a separate commit. During coding I saw this function was only called from within this file, so I moved it down with the other static functions and marked it as such.

By keeping it in a separate commit with a good message, I doubt we'll have any mistaken attribution (and more advanced git users would probably be using flags to ignore moves anyhow).

marcocitus commented on src/backend/distributed/shared_library_init.c:
```diff
@@ -1119,6 +1122,18 @@ RegisterCitusConfigVariables(void)
+		NULL,
+		&SetLocalWhitelistString,
+		"",
+		PGC_POSTMASTER,
```
Can we make this PGC_USERSET? Otherwise it isn't very useful in cloud.

jasonmp85 replied:
Sure; did it (pushing soon).

marcocitus commented on src/backend/distributed/transaction/transaction_management.c:
```diff
@@ -483,8 +495,12 @@ static void
+	SubXactState *state = linitial(activeSubXactStates);
+
+	Assert(state->subId == subId);
+	activeSetStmts = state->setLocalCmd;
```
I'm not sure I follow this logic. Should we not keep separate SET LOCAL lists for each subxact level?

jasonmp85 replied:
Writing an explanation in code comments, since I anticipated this would be the confusing part (I somewhat deviously/lazily like to see what code reviewers are confused by before explaining things in comments to see what I need to focus on).

jasonmp85 commented:
OK, @marcocitus… I've brushed up the documentation and (I think) addressed your question about the implementation. Basically:

  * During a transaction there's this `StringInfo` I append `SET LOCAL` stmts to
  * When we push into a subxact, the current value of that string is saved on the context stack and the string ref itself is reset to empty to keep track of subsequent `SET LOCAL` commands

Then:

  * Nodes already part of a transaction just get the `SET LOCAL` commands as the transaction progresses
  * Nodes coming into a transaction at a later point piggyback on the `SAVEPOINT` stacking behavior: instead of just sending them `SAVEPOINT`s, we send (pseudo-regex) `((SET LOCAL;)* SAVEPOINT)+(SET LOCAL;)*`

The `ROLLBACK` behavior of `SET LOCAL` ensures all this works in a stack-like way. The command strings live in the current transaction context so all that Just Works™, too.

**Let me know what i need to do to get this merged, i'd like to do so tomorrow and don't know what else is required now that all feedback is addressed**.

jasonmp85 commented:
@samay-sharma: Marco wanted to know what you thought about the use of a whitelist for specifying which GUC variables are subject to propagation. Do you think this works for customers? From what I've seen, the use-cases typically have specific variables they'd like propagated, so if they can identify and whitelist them they should be OK, IMO.

jasonmp85 commented:
> Looks fine.

Ok. But no PR approval? Ask I asked before: what’s needed for that?

> I'm wondering whether we should force router SELECTs to open a transaction block when there is a SET LOCAL.

Isn’t this scope creep? I’d really like to move on to something else unless this doesn’t address the issue it was written to address. Let me know.

jasonmp85 commented:
@marcocitus this is ready for a final review.

@samay-sharma (I think? Maybe it was you) mentioned offhandedly that an `enum` GUC would be more forward-looking, so I went with that instead of a `bool`; it'll eventually allow settings to permit propagation of all types of `SET` commands, but for now it's just `none` or `local` (default `none`).

The only things I thought might be worth including on a blacklist of "illegal" variables to propagate were `exit_on_error` (kills the session on any error, which would be bad for assumptions around connection reuse, I suppose) and `max_stack_depth` (we read this during initialization to eagerly attempt to allocate the stack (and fail fast) rather than letting the kernel do later so as it grows (and fail in a more roundabout/obtuse way)). I also threw in the `propagate_set_commands` GUC for good measure.

If you could look over the settings to see whether _you_ think anything needs banning, that would be good (I just read over them all using e.g. `SELECT name, short_desc  FROM pg_settings WHERE context IN ('user', 'superuser');`). Thanks!

marcocitus commented on src/backend/distributed/transaction/transaction_management.c:
```diff
@@ -46,8 +46,18 @@ XactModificationType XactModificationLevel = XACT_MODIFICATION_NONE;
-/* stack of active sub-transactions */
-static List *activeSubXacts = NIL;
+/* string to keep track of SET LOCAL statements executed within current subxact */
+/* set to NULL when pushing into new subxact or ending top xact */
```
slightly unusual comment formatting

jasonmp85 replied:
I think I added one line in one commit and the other in another; I'll combine.

marcocitus commented:
Note: I pushed a commit to fix the compilation issue, maybe squash it first.

jasonmp85 commented:
> This currently gives me a compile error

I really gotta figure out what's different between dev and CI; it's odd that the CI build doesn't similarly fail…

I don't write normal `for` loops very often, but when I do I _really_ dislike not being able to restrict the loop variable's scope to, well, that of the loop. We should address that someday. But for now I'll fix it.

jasonmp85 commented:
I think `session_preload_libraries` isn't going to do much given the user is already in a session (i.e. it's too late to change that at this point, no?)…

`application_name` is a good one. I'll add that.
